File: ruby/ocr-numbers/ocr_numbers.rb
# frozen_string_literal: true

module OcrNumbers
  CONVERSIONS = Hash.new("?").tap do |hash|
    hash[" _ | ||_|   "] = "0"
    hash["     |  |   "] = "1"
    hash[" _  _||_    "] = "2"
    hash[" _  _| _|   "] = "3"
    hash["   |_|  |   "] = "4"
    hash[" _ |_  _|   "] = "5"
    hash[" _ |_ |_|   "] = "6"
    hash[" _   |  |   "] = "7"
    hash[" _ |_||_|   "] = "8"
    hash[" _ |_| _|   "] = "9"
  end.freeze
  private_constant :CONVERSIONS
  HEIGHT = 4
  private_constant :HEIGHT
  WIDTH = 3
  private_constant :WIDTH

  module_function

  def convert(pipes)
    pipes = pipes.split("\n")
    raise ArgumentError unless valid_pipes?(pipes)
    pipes
      .map { |row| row.scan(/.{#{WIDTH}}/) }
      .transpose
      .map { |number_array| number_array.each_slice(HEIGHT).to_a }
      .transpose
      .map { |number_array| ocr_to_number(number_array) }
      .join(",")
  end

  def valid_pipes?(pipes)
    (pipes.length % HEIGHT).zero? &&
      pipes.all? { |pipe| (pipe.length % WIDTH).zero? }
  end
  private_class_method :valid_pipes?

  def ocr_to_number(number_array)
    number_array.map { |arr| CONVERSIONS[arr.join] }.join
  end
  private_class_method :ocr_to_number
end

module BookKeeping
  VERSION = 1
end

File: elixir/ocr-numbers/lib/ocr_numbers.ex
defmodule OcrNumbers do
  @conversions %{
    " _ | ||_|   " => "0",
    "     |  |   " => "1",
    " _  _||_    " => "2",
    " _  _| _|   " => "3",
    "   |_|  |   " => "4",
    " _ |_  _|   " => "5",
    " _ |_ |_|   " => "6",
    " _   |  |   " => "7",
    " _ |_||_|   " => "8",
    " _ |_| _|   " => "9"
  }
  @height 4
  @width 3
  @unrecognised "?"

  @doc """
  Given a 3 x 4 grid of pipes, underscores, and spaces, determine which number is represented, or
  whether it is garbled.
  """
  @spec convert([String.t()]) :: String.t()
  def convert(input) do
    with :ok <- check_height(input),
         :ok <- check_width(input) do
      {:ok, to_number(input)}
    end
  end

  defp check_height(input) do
    if valid_height?(input) do
      :ok
    else
      {:error, 'invalid line count'}
    end
  end

  defp valid_height?(input) do
    input
    |> length()
    |> rem(@height)
    |> Kernel.==(0)
  end

  defp check_width(input) do
    if Enum.all?(input, &valid_width?/1) do
      :ok
    else
      {:error, 'invalid column count'}
    end
  end

  defp valid_width?(input) do
    input
    |> String.length()
    |> rem(@width)
    |> Kernel.==(0)
  end

  defp to_number(input) do
    input
    |> Enum.chunk_every(@height)
    |> Enum.map(&convert_row/1)
    |> Enum.join(",")
  end

  defp convert_row(row) do
    row
    |> Enum.map(&split_by_digits/1)
    |> transpose()
    |> Enum.map(&Enum.join/1)
    |> Enum.map(&convert_to_number/1)
    |> Enum.join()
  end

  defp split_by_digits(line) do
    line
    |> String.graphemes()
    |> Enum.chunk_every(@width)
    |> Enum.map(&Enum.join/1)
  end

  defp transpose(row) do
    row
    |> Enum.zip()
    |> Enum.map(&Tuple.to_list/1)
  end

  def convert_to_number(ocr), do: Map.get(@conversions, ocr, @unrecognised)
end
